File: vandalizer-next/frontend/src/api/teams.ts
import { apiFetch } from './client'
import type { Team, TeamMember, TeamInvite } from '../types/user'

export function listTeams() {
  return apiFetch<Team[]>('/api/teams/')
}

export function createTeam(name: string) {
  return apiFetch<{ id: string; uuid: string; name: string }>('/api/teams/create', {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
}

export function updateTeamName(team_id: string, name: string) {
  return apiFetch<{ status: string; name: string }>('/api/teams/update_name', {
    method: 'PATCH',
    body: JSON.stringify({ team_id, name }),
  })
}

export function switchTeam(teamUuid: string) {
  return apiFetch<{ uuid: string; name: string }>(`/api/teams/switch/${teamUuid}`, {
    method: 'POST',
  })
}

export function getTeamMembers(teamUuid: string) {
  return apiFetch<TeamMember[]>(`/api/teams/${teamUuid}/members`)
}

export function getTeamInvites(teamUuid: string) {
  return apiFetch<TeamInvite[]>(`/api/teams/${teamUuid}/invites`)
}

export function inviteMember(team_id: string, email: string, role: string = 'member') {
  return apiFetch<{ token: string; email: string }>('/api/teams/invite', {
    method: 'POST',
    body: JSON.stringify({ team_id, email, role }),
  })
}

export function acceptInvite(token: string) {
  return apiFetch<{ uuid: string; name: string }>(`/api/teams/invite/accept/${token}`, {
    method: 'POST',
  })
}

export function changeMemberRole(team_id: string, user_id: string, role: string) {
  return apiFetch<{ ok: boolean }>('/api/teams/member/role', {
    method: 'POST',
    body: JSON.stringify({ team_id, user_id, role }),
  })
}

export function removeMember(team_id: string, user_id: string) {
  return apiFetch<{ ok: boolean }>('/api/teams/member/remove', {
    method: 'POST',
    body: JSON.stringify({ team_id, user_id }),
  })
}

File: vandalizer-next/backend/app/routers/teams.py
from fastapi import APIRouter, Depends, HTTPException

from app.dependencies import get_current_user
from app.models.user import User
from app.schemas.teams import (
    ChangeRoleRequest,
    CreateTeamRequest,
    InviteRequest,
    RemoveMemberRequest,
    UpdateTeamNameRequest,
)
from app.services import team_service

router = APIRouter()


@router.get("/")
async def list_teams(user: User = Depends(get_current_user)):
    """List all teams the user belongs to."""
    return await team_service.get_user_teams(user.user_id)


@router.get("/{team_uuid}/members")
async def list_members(team_uuid: str, user: User = Depends(get_current_user)):
    """List members of a team."""
    from app.models.team import Team

    team = await Team.find_one(Team.uuid == team_uuid)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    return await team_service.get_team_members(team.id)


@router.get("/{team_uuid}/invites")
async def list_invites(team_uuid: str, user: User = Depends(get_current_user)):
    """List pending invites for a team."""
    from app.models.team import Team

    team = await Team.find_one(Team.uuid == team_uuid)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    return await team_service.get_team_invites(team.id)


@router.post("/create")
async def create_team(
    body: CreateTeamRequest,
    user: User = Depends(get_current_user),
):
    team = await team_service.create_team(body.name, user.user_id)
    return {"id": str(team.id), "uuid": team.uuid, "name": team.name}


@router.patch("/update_name")
async def update_name(
    body: UpdateTeamNameRequest,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.update_team_name(
            body.team_id, body.name, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"status": "success", "name": team.name}


@router.post("/invite")
async def invite(
    body: InviteRequest,
    user: User = Depends(get_current_user),
):
    try:
        inv = await team_service.invite_member(
            body.team_id, body.email, body.role, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"token": inv.token, "email": inv.email}


@router.post("/invite/accept/{token}")
async def accept_invite(
    token: str,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.accept_invite(token, user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"uuid": team.uuid, "name": team.name}


@router.post("/switch/{team_uuid}")
async def switch_team(
    team_uuid: str,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.switch_team(team_uuid, user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"uuid": team.uuid, "name": team.name}


@router.post("/member/role")
async def change_role(
    body: ChangeRoleRequest,
    user: User = Depends(get_current_user),
):
    try:
        await team_service.change_role(
            body.team_id, body.user_id, body.role, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True}


@router.post("/member/remove")
async def remove_member(
    body: RemoveMemberRequest,
    user: User = Depends(get_current_user),
):
    try:
        await team_service.remove_member(
            body.team_id, body.user_id, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True}
